apps/passport-client/components/screens/LoginScreens/LoginScreen.tsx

```diff
@@ -126,6 +130,93 @@ export function LoginScreen(): JSX.Element {
+
+  /**
+   * Assuming we're in Chrome and an iframe, and we've successfully loaded an
+   * encryption key from local storage, try to use it to log in.
+   */
+  const tryToLogin = useCallback(
+    async (encryptionKey: string) => {
+      // Try to download and decrypt the storage
+      const storageRequest = await requestDownloadAndDecryptStorage(
+        appConfig.zupassServer,
+        encryptionKey
+      );
+      if (storageRequest.success) {
+        // Success, log in
+        dispatch({
+          type: "load-after-login",
+          storage: storageRequest.value,
+          encryptionKey
+        });
+      }
+    },
+    [dispatch]
+  );
+
+  /**
+   * This will only be called if we're in an iframe and Chrome.
+   */
+  const requestStorageAndLogIn = useCallback(async () => {
+    try {
+      // @ts-expect-error Chrome-only API
+      const handle: { localStorage: Storage } =
+        // @ts-expect-error Chrome-only API
+        await document.requestStorageAccess({ localStorage: true });
+
+      // Access granted, try reading the local storage
+      const encryptionKey = handle.localStorage.getItem("encryption_key");
+      if (encryptionKey) {
+        await tryToLogin(encryptionKey);
+      }
+    } catch (_e) {
+      // If the user rejected the storage access request, catch the exception
+      // but otherwise do nothing. The finally block will return the user to
+      // the regular login flow.
+    } finally {
+      setTryStorageAccess(false);
+      setTryStorageLogin(false);
+    }
+  }, [tryToLogin]);
+
+  useEffect(() => {
+    (async (): Promise<void> => {
+      // Are we in an iframe? If so, we might be able to skip requesting the
+      // user's email and password by retrieving their encryption key from the
+      // first-party local storage. Currently this only works on Chrome 125+.
+      const parser = new UAParser();
+      const browserName = parser.getBrowser().name;
+      const browserVersion = parser.getBrowser().version;
+      const isChrome125OrAbove =
+        browserName === "Chrome" &&
+        browserVersion &&
+        parseInt(browserVersion) >= 125;
+
+      if (window.parent !== window && isChrome125OrAbove) {
+        // Do we already have access?
+        const hasAccess = await document.hasStorageAccess();
+        if (!hasAccess) {
+          // No access, try requesting it interactively
+          // Setting this state will trigger the UI to show the "Connect to
+          // Zupass" button. To request storage access, the user must click
+          // the button and approve the dialog.
+          // Storage access requests must occur in response to a user action,
+          // so we can't request it automatically here and must wait for the
+          // user to click the button.
+          setTryStorageAccess(true);
+        } else {
+          // We have access, try logging in
+          // Show a spinner:
+          setTryStorageLogin(true);
```

Comment:
Also setTryStorageAccess(false), just to be sure your state-machine transitions are clear?
Or maybe these two booleans really should be an enum representing a state-machine directly.

Reply:
I've refactored the state machine into an enum with explicit states.